File: atividade_02/JavaScript/MaquinaDeLerArquivos.js
const fs = require('fs');
const path = require('path');
const LinkedList = require('./LinkedList.js');

class MaquinaDeLerArquivos {
    constructor() {
        this.lista = new LinkedList();
        this.qtdAcoes = 0;
        this.debugMode = false;
    }

    lerArquivo(caminho) {
        try {
            const data = fs.readFileSync(caminho, 'utf8');
            const linhas = data.split('\n').map(linha => linha.trim()).filter(linha => linha !== '');

            // Ler primeira linha com os valores iniciais da lista
            const valores = linhas[0].split(' ').map(Number);

            // Adiciona os valores na lista
            for (const valor of valores) {
                this.lista.addLast(valor);
            }

            // Ler quantidade de ações
            this.qtdAcoes = parseInt(linhas[1]);

            if (this.debugMode) {
                console.log();
                console.log("----- DEBUG MODE LIGADO -----");
                console.log();
                console.log(" __    __     ______     ______     __   ");
                console.log("/\\ \"-./  \\   /\\  __ \\   /\\  == \\   /\\ \\  ");
                console.log("\\ \\ \\-./\\ \\  \\ \\ \\/\\ \\  \\ \\  __<   \\ \\ \\ ");
                console.log(" \\ \\_\\ \\ \\_\\  \\ \\_____\\  \\ \\_\\ \\_\\  \\ \\_\\");
                console.log("  \\/_/  \\/_/   \\/_____/   \\/_/ /_/   \\/_/");
                console.log();
                console.log("Os prints com os índices só aparecerão no modo debug. O print das ações também.");
                console.log("[índice, valor]");
                console.log();
                console.log("Execução do código:");
                console.log();
            }

            // Executar as ações apenas da qtd de ações
            for (let i = 0; i < this.qtdAcoes; i++) {
                if (i + 2 < linhas.length) {
                    const linha = linhas[i + 2];
                    if (linha) {
                        this.executarAcao(linha);
                    }
                }
            }

        } catch (error) {
            if (error.code === 'ENOENT') {
                console.log(`Arquivo ${caminho} não encontrado`);
            } else {
                console.log(`Erro ao ler arquivo: ${error.message}`);
            }
        }
    }

    saveToFile(caminho) {
        try {
            let current = this.lista.head;
            const valores = [];
            while (current !== null) {
                valores.push(current.value.toString());
                current = current.next;
            }
            fs.writeFileSync(caminho, valores.join(' ') + '\n');
            console.log(`Lista salva em ${caminho}`);
            console.log();
        } catch (error) {
            console.log(`Erro ao salvar arquivo: ${error.message}`);
        }
    }

    executarAcao(linha) {
        const partes = linha.trim().split(' ');
        const acao = partes[0];

        switch (acao) {
            case 'A': // Adicionar
                const numero = parseInt(partes[1]);
                const posicao = parseInt(partes[2]);

                if (this.debugMode) {
                    try {
                        const valorAntes = this.lista.getValorPosicao(posicao);
                        console.log(`Adicionou ${numero} na posição ${posicao}, valor anterior dessa posição: ${valorAntes}`);
                        console.log();
                    } catch {
                        console.log(`Adicionou ${numero} na posição ${posicao} (posição estava vazia)`);
                        console.log();
                    }
                }
                this.lista.addAt(posicao, numero);
                break;

            case 'R': // Remover
                const valorRemover = parseInt(partes[1]);
                const posicaoRemovido = this.lista.removePrimeiroEncontrado(valorRemover);

                if (this.debugMode) {
                    if (posicaoRemovido !== -1) {
                        console.log(`Removeu ${valorRemover} da posição ${posicaoRemovido}`);
                        console.log();
                    }
                }
                break;

            case 'P': // Imprimir
                if (this.debugMode) {
                    this.lista.printListIndex();
                } else {
                    this.lista.printList();
                }
                break;

            case 'S': // SALVAR
                this.saveToFile("output.txt");
                break;
        }
    }
}

module.exports = MaquinaDeLerArquivos;

File: atividade_02/Python/maquina_de_ler_arquivos.py
from linked_list import LinkedList

class MaquinaDeLerArquivos:
    def __init__(self):
        self.lista = LinkedList()
        self.qtd_acoes = 0
        self.debug_mode = False

    def ler_arquivo(self, caminho):
        try:
            with open(caminho, 'r') as file:
                linhas = file.readlines()

            # Ler primeira linha com os valores iniciais da lista
            valores = list(map(int, linhas[0].strip().split()))

            # Adiciona os valores na lista
            for valor in valores:
                self.lista.add_last(valor)

            # Ler quantidade de ações
            self.qtd_acoes = int(linhas[1].strip())

            if self.debug_mode:
                print()
                print("----- DEBUG MODE LIGADO -----")
                print()
                print(" __    __     ______     ______     __   ")
                print("/\\ \"-./  \\   /\\  __ \\   /\\  == \\   /\\ \\  ")
                print("\\ \\ \\-./\\ \\  \\ \\ \\/\\ \\  \\ \\  __<   \\ \\ \\ ")
                print(" \\ \\_\\ \\ \\_\\  \\ \\_____\\  \\ \\_\\ \\_\\  \\ \\_\\")
                print("  \\/_/  \\/_/   \\/_____/   \\/_/ /_/   \\/_/")
                print()
                print("Os prints com os índices só aparecerão no modo debug. O print das ações também.")
                print("[índice, valor]")
                print()
                print("Execução do código:")
                print()

            # Executar as ações apenas da qtd de ações
            for i in range(self.qtd_acoes):
                if i + 2 < len(linhas):
                    linha = linhas[i + 2].strip()
                    if linha:
                        self.executar_acao(linha)

        except FileNotFoundError:
            print(f"Arquivo {caminho} não encontrado")
        except Exception as e:
            print(f"Erro ao ler arquivo: {e}")

    def save_to_file(self, caminho):
        try:
            with open(caminho, 'w') as file:
                current = self.lista.head
                valores = []
                while current is not None:
                    valores.append(str(current.value))
                    current = current.next
                file.write(' '.join(valores) + '\n')
            print(f"Lista salva em {caminho}")
            print()
        except Exception as e:
            print(f"Erro ao salvar arquivo: {e}")


    def executar_acao(self, linha):
        partes = linha.strip().split()
        acao = partes[0]

        if acao == "A":  # Adicionar
            numero = int(partes[1])
            posicao = int(partes[2])

            if self.debug_mode:
                try:
                    valor_antes = self.lista.get_valor_posicao(posicao)
                    print(f"Adicionou {numero} na posição {posicao}, valor anterior dessa posição: {valor_antes}")
                    print()
                except:
                    print(f"Adicionou {numero} na posição {posicao} (posição estava vazia)")
                    print()

            self.lista.add_at(posicao, numero)

        elif acao == "R":  # Remover
            valor_remover = int(partes[1])
            posicao_removido = self.lista.remove_primeiro_encontrado(valor_remover)

            if self.debug_mode:
                if posicao_removido != -1:
                    print(f"Removeu {valor_remover} da posição {posicao_removido}")
                    print()

        elif acao == "P":  # Imprimir
            if self.debug_mode:
                self.lista.print_list_index()
            else:
                self.lista.print_list()

        elif acao == "S":  # SALVAR
            self.save_to_file("output.txt")

